common/src/main/java/io/netty/util/DomainWildcardMappingBuilder.java

```diff
@@ -0,0 +1,160 @@
+     * @param defaultValue    the default value for {@link Mapping#map(Object)} to return
+     *                        when nothing matches the input
+     */
+    public DomainWildcardMappingBuilder(int initialCapacity, V defaultValue) {
+        this.defaultValue = checkNotNull(defaultValue, "defaultValue");
+        map = new LinkedHashMap<String, V>(initialCapacity);
+    }
+
+    /**
+     * Adds a mapping that maps the specified (optionally wildcard) host name to the specified output value.
+     * {@code null} values are forbidden for both hostnames and values.
+     * <p>
+     * <a href="https://tools.ietf.org/search/rfc6125#section-6.4">DNS wildcard</a> is supported as hostname. The
+     * wildcard will only match one sub-domain deep and only when wildcard is used as the most-left label.
+     *
+     * For example:
+     *
+     * <p>
+     *  *.netty.io will match xyz.netty.io but NOT abc.xyz.netty.io
+     * </p>
+     *
+     * @param hostname the host name (optionally wildcard)
+     * @param output   the output value that will be returned by {@link Mapping#map(Object)}
+     *                 when the specified host name matches the specified input host name
+     */
+    public DomainWildcardMappingBuilder<V> add(String hostname, V output) {
+        map.put(normalizeHostName(hostname),
+                checkNotNull(output, "output"));
+        return this;
+    }
+
+    private String normalizeHostName(String hostname) {
+        checkNotNull(hostname, "hostname");
+        if (hostname.isEmpty() || hostname.charAt(0) == '.') {
+            throw new IllegalArgumentException("Hostname '" + hostname + "'not valid");
+        }
+        hostname = ImmutableDomainWildcardMapping.normalize(checkNotNull(hostname, "hostname"));
+        if (hostname.charAt(0) == '*') {
+            if (hostname.length() < 3 || hostname.charAt(1) != '.') {
+                throw new IllegalArgumentException("Wildcard Hostname '" + hostname + "'not valid");
+            }
+            return hostname.substring(1);
+        }
+        return hostname;
+    }
+    /**
+     * Creates a new instance of an immutable {@link Mapping}.
+     *
+     * @return new {@link Mapping} instance
+     */
+    public Mapping<String, V> build() {
+        return new ImmutableDomainWildcardMapping<V>(defaultValue, new LinkedHashMap<String, V>(map));
+    }
+
+    private static final class ImmutableDomainWildcardMapping<V> implements Mapping<String, V> {
+        private static final String REPR_HEADER = "ImmutableDomainWildcardMapping(default: ";
+        private static final String REPR_MAP_OPENING = ", map: ";
+        private static final String REPR_MAP_CLOSING = ")";
+
+        private final V defaultValue;
+        private final Map<String, V> map;
+
+        ImmutableDomainWildcardMapping(V defaultValue, Map<String, V> map) {
+            this.defaultValue = defaultValue;
+            this.map = Collections.unmodifiableMap(map);
```

Comment:
`DomainWildcardMappingBuilder::build` is alredy creating a fresh new map that would be owned by `ImmutableDomainWildcardMapping`: if you let the copy of `DomainWildcardMappingBuilder::map` to happen inside `new ImmutableDomainWildcardMapping`, you could save using an `unmodifiableMap` and the pointer chasing introduced by it

Reply:
@franz1981 good point...